// @chainlink/solana-sdk/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OCR2Feed = exports.CHAINLINK_STORE_PROGRAM_ID = exports.CHAINLINK_AGGREGATOR_PROGRAM_ID = void 0;
const tslib_1 = require("tslib");
const anchor = (0, tslib_1.__importStar)(require("@project-serum/anchor"));
const anchor_1 = require("@project-serum/anchor");
const web3_js_1 = require("@solana/web3.js");
exports.CHAINLINK_AGGREGATOR_PROGRAM_ID = new web3_js_1.PublicKey("cjg3oHmg9uuPsP8D6g29NWvhySJkdYdAo9D25PRbKXJ");
exports.CHAINLINK_STORE_PROGRAM_ID = new web3_js_1.PublicKey("HEvSKofvBgfaexv23kMabbYqxasxU3mQ4ibBMEmJWHny");
class OCR2Feed {
    constructor(aggregatorProgram, provider) {
        this.aggregatorProgram = aggregatorProgram;
        this.provider = provider;
        this._parser = new anchor.EventParser(aggregatorProgram.programId, aggregatorProgram.coder);
    }
    static async load(programID = exports.CHAINLINK_AGGREGATOR_PROGRAM_ID, provider = anchor.AnchorProvider.env()) {
        const aggregatorProgram = await anchor.Program.at(programID, provider);
        return new OCR2Feed(aggregatorProgram, provider);
    }
    onRound(feed, callback) {
        return this.provider.connection.onLogs(feed, (event, ctx) => {
            this._parser.parseLogs(event.logs, (log) => {
                if (log.name != "NewTransmission") {
                    return;
                }
                let parsed = OCR2Feed.parseLog(log);
                parsed.feed = feed;
                parsed.slot = ctx.slot;
                callback(parsed);
            });
        });
    }
    async removeListener(listener) {
        return this.provider.connection.removeOnLogsListener(listener);
    }
    static parseLog(log) {
        if (!log || !log.data)
            return null;
        let answer;
        if (log.data.answer)
            answer = new anchor_1.BN(log.data.answer, 10, "le");
        let roundId;
        if (log.data.roundId)
            roundId = log.data.roundId;
        let observationsTS;
        if (log.data.observationsTimestamp)
            observationsTS = new Date(log.data.observationsTimestamp * 1000);
        return {
            answer: answer,
            roundId: roundId,
            observationsTS: observationsTS,
        };
    }
}
exports.OCR2Feed = OCR2Feed;
